src/dune_manager/dune_manager.ml

```diff
@@ -420,13 +420,20 @@ let daemon ~root ~config started =
   let log_file = Path.relative root "log" in
   Log.init ~file:(This log_file) ();
   let manager = make ~root ~config () in
-  let handler _ =
-    Log.info "caught signal";
+  (* Event blocks signals when waiting. Use a separate thread to catch signals. *)
+  let signal_handler s =
+    Log.infof "caught signal %i, exiting" s;
     ignore (Thread.create stop manager)
+  and signals = [ Sys.sigint; Sys.sigterm ] in
+  let rec signals_handler () =
+    ignore (Thread.sigmask Unix.SIG_UNBLOCK signals);
+    signal_handler (Thread.wait_signal signals);
+    signals_handler ()
   in
-  (* Unfortunately it seems using Event prevents signals from working :-( *)
-  Sys.set_signal Sys.sigint (Sys.Signal_handle handler);
-  Sys.set_signal Sys.sigterm (Sys.Signal_handle handler);
+  Sys.set_signal Sys.sigint Sys.Signal_ignore;
+  Sys.set_signal Sys.sigterm Sys.Signal_ignore;
+  ignore (Thread.create signals_handler ());
+  ignore (Thread.sigmask Unix.SIG_BLOCK signals);
```

Comment:
We do something similar in the src/dune/scheduler.ml. The pattern should be:

1. block the signals before starting any thread
2. call wait_signal in a loop in the signal handler thread

there is no need to unblock signals in the signal handler thread and there is no need to set the signal disposition to ignore.